Refactor code/evaluators to allow implementation of DSL based on enhanced functionality

trulede opened the pull request:
Refactor code/evaluators to allow implementation of DSL based on enhanced functionality which was originally located in the Python Evaluator. The PythonEvaluator now only deals with Python Code, and the EnhancedEvaluator concerns itself with various behaviours which are extensions to the base Evaluator Class.

Its possible to implement an Evaluator against either the Evaluator or EnhancedEvaluator interface. With this rework I will try both (or at least the later) as a basis for a DSL implementation based on Python Lark. 

Longer term I would offer the suggestion of converting the EnhancedEvaluator into a Mixin Class so that it becomes easier to expand the Evaluator behaviour without affecting the existing API interfaces. If I see an opportunity to try this, while developing the DSL, I will give it a try.

AlexandreDecan commented:
Thanks for this PR! I'll try to review it asap, but I'm afraid I won't have time before next Monday or next Tuesday.

AlexandreDecan commented on sismic/code/evaluator.py:
```diff
@@ -88,9 +88,7 @@ def evaluate_guard(self, transition: Transition, event: Optional[Event]=None) ->
-        if transition.guard:
-            return self._evaluate_code(transition.guard, additional_context={'event': event})
-        return None
+        return self._evaluate_code(getattr(transition, 'guard', None), additional_context={'event': event})
```
This is not equivalent if `transition.guard` is `None`. In the first case, `_evaluate_code` won't be called, where in the second case, it will be called with `None`. I don't have time to check right now, but either you should not remove the conditional statement, or you should set default value to the empty string?

trulede replied:
The Python Evaluator **was** doing this:
```
return self._evaluate_code(getattr(transition, 'guard', None),  ...)
```
where the code evaluates to either a string of code or None. That was then handled in the _evaluate_code/_execute_code block with:
```
if code is None:
    return []
```

I thought the implementation in Python Evaluator was more succinct and therefore adopted it to both the Evaluator and EnhancedEvaluator classes so that implementations would be consistent. Perhaps what really needs to happen is to update the abstract function definitions for _evaluate_code() and _execute_code() along these lines

```
@abc.abstractmethod
 def _execute_code(self, code: Optional[str]=None, *, additional_context: Mapping[str, Any]=None) -> List[Event]:
        foo = bar
```
If I have that bit correct ... **Optional[str]=None** ... I will test it later anyway, just to see.

I think None is better than an empty string, since an empty string is valid code, and None indicates _no_ code at all ... but I don't mind either way :-)  I've got the example Lark DSL wired up with Sismic via this mechanism, it works! I hit this problem too ... but what you had in the Python Evaluator was fine by me.

AlexandreDecan replied:
You're right ;) Let's continue with `Optional[str]` in both `_execute_code` and `_evaluate_code`.

AlexandreDecan commented on sismic/code/evaluator.py:
```diff
@@ -101,10 +99,7 @@ def execute_action(self, transition: Transition, event: Optional[Event]=None) ->
-            return self._execute_code(transition.action, additional_context={'event': event})
-        else:
-            return []
+        return self._execute_code(getattr(transition, 'action', None), additional_context={'event': event})
```
Probably the same, here.

AlexandreDecan commented on sismic/code/evaluator.py:
```diff
@@ -114,11 +109,7 @@ def execute_on_entry(self, state: StateMixin) -> List[Event]:
-            return self._execute_code(code)
-        else:
-            return []
+        return self._execute_code(getattr(state, 'on_entry', None))
```
Probably the same here.

AlexandreDecan commented on sismic/code/evaluator.py:
```diff
@@ -128,11 +119,7 @@ def execute_on_exit(self, state: StateMixin) -> List[Event]:
-            return self._execute_code(code)
-        else:
-            return []
+        return self._execute_code(getattr(state, 'on_exit', None))
```
Probably the same here ;)

AlexandreDecan commented:
... however, I quickly looked at the PR, and I think that too many things were moved to `EnhancedEvaluator' :-) I agree with you to go for mixins instead. I've started a review, but it won't be complete ;)

AlexandreDecan commented on sismic/code/python.py:
```diff
@@ -54,43 +20,10 @@ def notify(name, **kwargs):
     """
     A code evaluator that understands Python.
 
-    Depending on the method that is called, the context can expose additional values:
```
Docstring should remain, notably because it is currently used by Sphinx to generate "official" documentation.

AlexandreDecan replied:
One possibility is to explain in the documentation that we have two base evaluators, a very simple one that does nothing, and another one that supports several features (time, active(), send(), ....), and one concrete implementation of that one (EnhancedEvaluator should be an ABCMeta then) that supports Python code.

trulede replied:
I moved the doc (to EnhancedEvaluator) but wanted to wait for your feedback regarding creating Mixin's before doing anymore with it. If Sphinx will not pick it up where it is, then OK, drop it for now.

trulede commented:
As additional info, this is how I used the Evaluator today:

```
class LarkDslEvaluator(EnhancedEvaluator):
    def __init__(self, interpreter=None, *, initial_context=None):
        super().__init__(interpreter, initial_context=initial_context)

        self._parser = initial_context['dsl_parser']
        self._engine = initial_context['dsl_engine']

    @property
    def context(self):
        return dict()

    def _evaluate_code(self, code: str, *, additional_context: Mapping=None) -> bool:
        if code is None:
            return True

        parse_tree = self._parser.parse(code)
        for inst in parse_tree.children:
            self._engine.run_instruction(inst)

        return True

    def _execute_code(self, code: str, *, additional_context: Mapping=None) -> List[Event]:
        if code is None:
            return []

        parse_tree = self._parser.parse(code)
        for inst in parse_tree.children:
            self._engine.run_instruction(inst)

        return []
```

Not really sure what initial_context was intended for ... I used it to pass some Objects into the Evaluator - they have to come from somewhere - but perhaps that is a misuse?

This simple example works when inheriting from either Evaluator or EnhancedEvaluator. Its pretty neat!

AlexandreDecan commented:
The `initial_context` of an `Evaluator` is set by the `Interpreter`, when an `initial_context` is provided upon creation. The purpose of this was to have an easy way to provide "default values" or "parameters" when a statechart is interpreted.

trulede commented:
Should be fine now.

AlexandreDecan commented:
Thanks. 

I'm not (yet) fully convinced this is the best approach to make the "goodies" of the (old) Python evaluator available. I really liked the idea of having mixins but I don't know how we can easily implement that (given that some elements are context dependent, and other ones need access to some "callback" to change their state). 

I wonder if we can achieve the EnhancedEvaluator directly from the PythonEvaluator class, e.g. by defining two new methods (_execute_code_with_context and _evaluate_code_with_context) that are called by _execute_code and _evaluate_code and simply evaluate/execute Python code, where the later enrich the context. This way, one can "simply" subclass PythonEvaluator and override _execute_code_with_context and _evaluate_code_with_context.

Currently, my main concern is that EnhancedEvaluator does not really provide something useful on its own and is very specific to your use case.

trulede commented:
I'm interested in using the additional context too! This is part of the problem, those things (time, events, context ...) are actually pretty useful. They are, I guess, "behaviours" and properties of the Statechart rather than Python specific - or do I understand the intention incorrectly?

I've used both the Evaluator and EnhancedEvaluator classes, it makes no sense to have both ... and the Evaluator class is not much use to me - since it misses all the interesting **and necessary** behaviours - and I can switch to the EnhancedEvaluator with no code changes other than the class name ... which is a big hint ;) 

What I would do myself is one of these two things (possibly a combination of both):

1. Merge the Evaluator and EnhancedEvaluator to a single Evaluator, so that all behaviour is in the base class, and all language specific processing is in the derived class.
2. Rework the EnhancedEvaluator into Mixin's.

Let me know if you are willing to take either or both of those paths, I can spend some more time on it ... I think some things should go in the Evaluator class (time & events in particular), and others would be better suited to a Mixin (the FrozenContext perhaps?).

AlexandreDecan commented:
I would like to keep `Evaluator` as simple as possible, to ease the implementation of new language interpreters (without having to worry about all the "goodies"). Right now, I think it's better to slightly adapt the `PythonEvaluator` so you can easily subclass and use it for your DSL, then I'll release Sismic 1.4.0, then we can discuss on how to provide these goodies (time, sent, old, ...) for reuse (e.g. via mixins). 

If you agree on this, I would suggest to move the "Python execution/evaluation part" of `_evaluate_code` and `_execute_code`of the `PythonEvaluator` to something like `_evaluate_code_with_context` and `_execute_code_with_context`, and to call these to methods from `_evaluate_code` and `_execute_code` (these two being responsible of providing the context containing the goodies that were not added by other methods e.g. `evaluate_preconditions`, etc.).

I think this first step will be sufficient for you to implement your own DSL on top of the Python evaluator, right?

trulede commented:
Lets put this to one side for now.

It seems like I can convert EnhancedEvaluator to a Mixin without too much effort 😉 so I will implement against the original Evaluator class and see how far I get with the behaviours I've extracted so far.

AlexandreDecan commented:
If we go for a mixin-based implementations of the goodies, it could be interesting to split them into multiple mixins (e.g. one to support after/idle/time, one to support sent/received, one to support __old__, one to support send/notify and one to support active for example). I don't know how much work it represents. 

For most of them, it should be quite easy (because they do not really rely on something external), but for some of them, it could be difficult to provide them in an elegant way (e.g. after/idle that need to "intercept" entry/exit/transition). 

That's why as a first step, I suggest to slightly alter PythonEvaluator so you can already extend it for your DSL-based evaluator, and then we can start discussing about the best way to support all the goodies with a mixin-based approach.

trulede commented:
Thanks, but its not necessary. I have the DSL working already :smile: , against the EnhancedEvaluator interface (now a Mixin to the Evaluator) ... and all of that was extracted from the Python Evaluator anyway. So I'm good.

I've put the original refactor here: https://github.com/trulede/sismic/tree/evaluator_refactor
And I'm working on the Mixin here: https://github.com/trulede/sismic/tree/working  and I should probably add the DSL Code ... if I can find where I put it :blush:

For the Mixin, I found the following article interesting, I think I would try something using a variation of this technique when it comes to splitting the thing up:
https://threadthehaystack.wordpress.com/2016/09/15/inheritance-super-and-mixins-in-python/

I've also done solved a similar problem to the entry/exit/transition using a decorator.

trulede commented:
Some experimentation  ... this is how I would handle the so called Cooperative Multiple Inheritance(Mixin) for Evaluator capabilities:

```
class I(metaclass=abc.ABCMeta):
	@abc.abstractmethod
	def foo(self):
		pass # NOP (no raise!)

class A(I):
	def foo(self):
		print("bar A")
		super().foo()
		
class B(I):
	def foo(self):
		print("bar B")
		super().foo()

class C(B, A):
	def __init__(self):
		super().__init__()

c = C()
>>> c.foo()
bar B
bar A
```

That's done the diamond with the Abstract Interface being a NOP. Its not perfect, but almost there.

In the case of the Code Evaluator, you might have:

- EvaluatorInterface.py - abstract class/interface definition.
- Evaluator.py - inherits EvaluatorInterface with existing implementation. Essentially a Mixin
- EvaluatorMixin(s) - inherits EvaluatorInterface, several classes, for each behaviour/capability.
- PythonEvaluator - inherits several EvaluatorMixins.

Still need to see how it works with decorators, however, so far I think I will use this technique. Still need to see about the Interface definition ... there might be a better way for that. In any case, it works nicely.

I've done something similar with configuration management, where a Base Class implements a configuration store with some additional capability (name resolution) and then derived Mixin classes are able to use that configuration store, and add to its capabilities (anonymous functions IIRC). The net result is to make some advanced capabilities available with out really noticing it.

AlexandreDecan commented:
That's interesting. However, most methods in an "Evaluator with goodies" are expected to do something *and* to return something (e.g. `execute_on_entry` has to execute code in an Evaluator, and to reset entry time/memory in a mixin). I think it will be quite confusing if we "mix" these. 

I'm currently working on how to implement a kind of `ContextProvider` that is independent (but used by) an evaluator. For these context providers to be independent of the evaluator, I need a way to "notify" them about what happens during the execution (e.g. a state was entered). It's very close to what is currently achieved with property statecharts, but I don't want to mix these concepts either. 

I think I'll try to extract the "meta-event monitoring" part of the property statecharts so it can be used by both property statechart *and* arbitrary listener.

trulede commented:
No, its not confusing :wink: People implementing their own Language support are going to know what they are doing.

Anyway, rather than explain why using super() will work .... I just implemented it instead 😎  

Take a look here, https://github.com/trulede/sismic/tree/evaluator_refactor , and let me know if you are interested in a Pull Request to bring in this functionality (with further refinements and doc). 

This is the diff, but its built on top of the other refactoring (and might not be clear to you): https://github.com/trulede/sismic/commit/372e2464af2c4960f06c46fa546550d3a690b11f

I've split the idle/after functionality into a separate Mixin Class to illustrate how it works.

- **Evaluator** - the basic implementation and interface definition.
- **EnhancedMixin** - a mixin with various functionality.
- **IdleAfterMixin** - a mixin with time based functionality.
- **PythonEvaluator** - an implementation of Python Language support which uses the above two mixins.

AlexandreDecan commented:
I've just discovered your answer, after having tried to implement the time/event related predicates apart of ``PythonEvaluator``. The implementation is based on context providers. A context provider exposes functions and variables that could be useful in an evaluator (e.g. after, idle, etc.). Context providers are based on the new binding system of an ``Interpreter``. Without going into details, this new binding system allows a listener to be notified when events (internal or meta- events) are sent by a statechart. The context providers rely on these meta-events to "compute" the truth value of the predicates. Notice that property statecharts now also rely on this new binding system to work. Probably this new system will be useful for you as well, as it allows to implement meta-events listeners without having to implement them as property statecharts.

AlexandreDecan commented:
I'll wait for your feedback about how "goodies" can be reused in code evaluator, and about the new binding system before releasing Sismic 1.4.0. I suggest to open new issues (if there are some issues ;-)). Thank you!;)
